architecture-docs/adr-002.md

```diff
@@ -0,0 +1,490 @@
+validators can be added or removed based on their network operations).
+
+### Problems with current design
+
+1. Our current design does not delete validator information from `ValidatorState` when a validator either unbonds their
+   funds or gets removed from validator set because of some punishment. The reason for not deleting this information is
+   because it is still required for computing `SlashRatio` while slashing a punished validator. Because of this,
+   validators cannot rejoin validator set (because their information already exists and we cannot add a duplicate
+   entry).
+1. During rewards calculation, we calculate `total_stake` based of the active validator at the time of rewards
+   distribution. This does not take into account all the validator set changes that took place in current rewards
+   period. For example, when a validator unbonds or deposits funds just before rewards distribution (let us assume that
+   they unbonded their funds at 23rd hour), their stake is not considered in rewards calculation.
+
+Besides this, validator set, rewards pool and punishments are three different components of our system and should
+maintain their own data instead of relying on `ValidatorState` for information. There should be a specified
+interface/behavior for interaction between these components.
+
+## Decision
+
+This document proposes new design for Dynamic Validator Set, Punishments and Rewards Pool. At the end, it specifies the
+interaction between all three components. 
+
+### Dynamic Validator Set
+
+`Validator` struct holds all the information related to a validator/council node.
+
+```rust
+/// A validator in Crypto.com Chain (this data structure can be merged with existing `CouncilNode`)
+pub struct Validator {
+    /// Name of validator
+    pub name: String,
+    /// Staking address of validator's staking account
+    pub staking_address: StakedStateAddress,
+    /// Tendermint's public key of validator
+    pub tendermint_pub_key: TendermintPubKey,
+    /// Current voting power of validator
+    pub voting_power: TendermintVotePower,
+    /// Optional security contact information
+    pub security_contact: Option<String>,
+}
+```
+
+Current validator set is maintained by `ValidatorState` struct. It exposes functions to add, remove and update current
+validator set. Besides this, `ValidatorState` also exposes function to calculate total voting power of current validator
+set which is used to calculate slash ratio while slashing a validator.
+
+```rust
+/// Runtime state of validator set
+pub struct ValidatorState {
+    /// All the validators corresponding to their staking address
+    validators: BTreeMap<StakedStateAddress, Validator>,
+    /// Staking address of validators based on their tendermint validator address (can be used to find validator
+    /// information from tendermint validator address)
+    tendermint_addresses: BTreeMap<TendermintValidatorAddress, StakedStateAddress>,
+}
```

Comment:
How do we calculate validator power updates(`Vec<(ValidatorPubKey, VotingPower)>`) based on this structure?

Reply:
Can you explain a bit more? Why do you want to calculate this information? It is already available to us in `end_block` as `power_changed_in_block`.